lib/maxmind/db/decoder.rb

```diff
@@ -0,0 +1,246 @@
+      end
+
+      def decode_uint128(size, offset)
+        if size == 0
+          return 0, offset
+        end
+
+        buf = @io.read(offset, size)
+
+        if size <= 8
+          buf = buf.rjust(8, "\x00".freeze)
+          return buf.unpack('Q>'.freeze)[0], offset + size
+        end
+
+        a_bytes = buf[0...-8].rjust(8, "\x00".freeze)
+        b_bytes = buf[-8...buf.length]
+        a = a_bytes.unpack('Q>'.freeze)[0]
+        b = b_bytes.unpack('Q>'.freeze)[0]
+        a <<= 64
+        return a | b, offset + size
+      end
+
+      def decode_map(size, offset)
+        container = {}
+        size.times do
+          key, offset = decode(offset)
+          value, offset = decode(offset)
+          container[key] = value
+        end
+        return container, offset
+      end
+
+      def decode_pointer(size, offset)
+        pointer_size = size >> 3
+
+        if pointer_size == 0
+          new_offset = offset + 1
+          buf = (size & 0x7).chr << @io.read(offset, 1)
+          pointer = buf.unpack('n'.freeze)[0] + @pointer_base
+        elsif pointer_size == 1
+          new_offset = offset + 2
+          buf = "\x00".freeze.b << (size & 0x7).chr << @io.read(offset, 2)
+          pointer = buf.unpack('N'.freeze)[0] + 2048 + @pointer_base
+        elsif pointer_size == 2
+          new_offset = offset + 3
+          buf = (size & 0x7).chr << @io.read(offset, 3)
+          pointer = buf.unpack('N'.freeze)[0] + 526336 + @pointer_base
+        else
+          new_offset = offset + 4
+          buf = @io.read(offset, 4)
+          pointer = buf.unpack('N'.freeze)[0] + @pointer_base
+        end
+
+        if @pointer_test
+          return pointer, new_offset
+        end
+
+        value, _ = decode(pointer)
+        return value, new_offset
+      end
+
+      def decode_utf8_string(size, offset)
+        new_offset = offset + size
+        buf = @io.read(offset, size)
+        buf.force_encoding(Encoding::UTF_8)
+        # We could check it's valid UTF-8 with `valid_encoding?', but for
+        # performance I do not.
+        return buf, new_offset
+      end
+
+      TYPE_DECODER = {
+        1  => :decode_pointer,
+        2  => :decode_utf8_string,
+        3  => :decode_double,
+        4  => :decode_bytes,
+        5  => :decode_uint16,
+        6  => :decode_uint32,
+        7  => :decode_map,
+        8  => :decode_int32,
+        9  => :decode_uint64,
+        10 => :decode_uint128,
+        11 => :decode_array,
+        14 => :decode_boolean,
+        15 => :decode_float,
+      }.freeze
+      private_constant :TYPE_DECODER
+
+      public
+
+      # Decode a section of the data section starting at +offset+.
+      #
+      # +offset+ is the location of the data structure to decode.
+      #
+      # Returns an array where the first element is the decoded value and the
+      # second is the offset after decoding it.
+      #
+      # Throws an exception if there is an error.
```

Comment:
I do wonder about this `throw` vs `raise again. You might know better than I do.